### PR TITLE
Adds Sentry breadcrumb feature

### PR DESCRIPTION
**Pull request by WangleLine**
Still need to test it out in production, but I'd like to hear what you have to say before I do that

Addresses #21

### CODE REVIEW

File: scripts/SnitchSentryBreadcrumbsGet/SnitchSentryBreadcrumbsGet.gml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/// Poop
```
Comment (WangleLine) [inline scripts/SnitchSentryBreadcrumbsGet/SnitchSentryBreadcrumbsGet.gml]: oops

File: scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+
+/// Adds a breadcrumb to an array of breadcrumbs which will be send to Sentry whenever the game crashes
+///
+/// @param string  human-readable message
+/// @param struct  custom struct with optional extra data
+
+function SnitchSentryBreadcrumb(_message,_data=undefined)
+{
+	if SNITCH_BREADCRUMBS_MAX == 0 return;
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml]: `<= 0` please (also needs brackets around the condition)

File: scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+	var _struct = {
+		message : _message
+	}
+	
+	// Add optional data if its defined
+	if _data != undefined
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml]: Add brackets around if-statement conditions i.e. `if (_data != undefined)`

File: scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+
+/// Adds a custom breadcrumb struct to an array of breadcrumbs which will be send to Sentry whenever the game crashes
+/// Read the sentry event payload documentation to find out more about what kind of data this struct can contain
+
+function SnitchSentryBreadcrumbExt(_struct)
+{
+	if SNITCH_BREADCRUMBS_MAX == 0 return;
+	
+	static array = __SnitchState().__breadcrumbArray;
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml]: Underscore prefix for this array i.e. `static _array = `

File: scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	
+	array_push(array,_struct);
+	
+	//Automatically add timestamp if struct doesn't have this member
+	if variable_struct_exists(_struct,"timestamp") == false
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml]: `if (not variable_struct_exists(...))` please

File: scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	{
+		_struct.timestamp = SnitchFormatTimestamp(date_current_datetime());
+	}
+	
+	//Prunes oldest breadcrumb if too many exist
+	if array_length(array) > SNITCH_BREADCRUMBS_MAX
+	{
+		array_delete(array,0,1);
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml]: Please space out parameters i.e. `array_delete(_array, 0, 1);`

File: scripts/__SnitchConfigServices/__SnitchConfigServices.gml [hunk 7]
```diff
@@ -60,4 +60,8 @@
 //   N.B. HTTP warnings/failures will always be reported
 #macro SNITCH_OUTPUT_HTTP_SUCCESS  (SNITCH_RUNNING_FROM_IDE)
 
+//How many breadcrumbs to keep
```
Comment (JujuAdams) [inline scripts/__SnitchConfigServices/__SnitchConfigServices.gml:63]: Link to `https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/` here too please

File: scripts/__SnitchInitialize/__SnitchInitialize.gml [hunk 8]
```diff
@@ -65,7 +65,9 @@ function __SnitchInitialize()
         __GAPlatform  = "";
         __GAOSVersion = "";
         __GAGameMakerVersion = "gamemaker " + string_delete(string_delete(GM_runtime_version, 5, 1), 1, 3);
-    
+		
+		//Breadcrumbs
```
Comment (JujuAdams) [inline scripts/__SnitchInitialize/__SnitchInitialize.gml:69]: Spaces not tabs in the future please (I'll accept w/o this change and I can fix it later)

File: scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+///
+/// @param string  human-readable message
+/// @param struct  custom struct with optional extra data
+
+function SnitchSentryBreadcrumb(_message,_data=undefined)
+{
+	if SNITCH_BREADCRUMBS_MAX == 0 return;
+
+	var _struct = {
+		message : _message
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml]: Needs stringification here i.e. `message: string(_message)`

File: scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+// https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+
+/// Adds a breadcrumb to an array of breadcrumbs which will be send to Sentry whenever the game crashes
+///
+/// @param string  human-readable message
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumb/SnitchSentryBreadcrumb.gml]: Please fix capitalisation of parameter description

File: scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml [hunk 11]
```diff
@@ -0,0 +1,25 @@
+// https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+
+/// Adds a custom breadcrumb struct to an array of breadcrumbs which will be send to Sentry whenever the game crashes
+/// Read the sentry event payload documentation to find out more about what kind of data this struct can contain
+
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumbExt/SnitchSentryBreadcrumbExt.gml:7]: Missing parameter description

File: scripts/SnitchSentryBreadcrumbsGet/SnitchSentryBreadcrumbsGet.gml [hunk 12]
```diff
@@ -0,0 +1,6 @@
+/// Returns all currently collected breadcrumbs
```
Comment (JujuAdams) [inline scripts/SnitchSentryBreadcrumbsGet/SnitchSentryBreadcrumbsGet.gml:3]: Needs comment that this function will always return an empty array if `SNITCH_BREADCRUMBS_MAX` is less than or equal to zero

**Comment by WangleLine** [comment]
Addressed everything~

**Comment by JujuAdams** [comment]
Ah shit forgot the check where the merge was going  _scream_

**Pull request merged**